### PR TITLE
Add temporary code to create reference table 3

### PR DESCRIPTION
**Pull request by MrPowers**
This is a DRAFT pull request to add a reference table that includes append and overwrite transactions.  The reference table can be created with this command: `python -m dat.main write-generated-reference-table-3`.

Will need to figure out how to adapt this code to fit the existing framework.  Some specific modifications we'll probably need to the existing framework:

* ability to write out multiple expected files to test Delta Lake table contents at specific versions
* optionally run `repartition(1)` before writing, so fewer Parquet files are generated

### CODE REVIEW

File: dat/main.py [hunk 1]
```diff
@@ -65,6 +65,11 @@ def write_generated_reference_tables(table_names, output_path):
     logging.info('Generated reference tables successfully written')
 
 
+@click.command()
+def write_generated_reference_table_3():
+    create_reference_tables.create_reference_table_3()
```
Comment (scottsand-db) [inline dat/main.py:70]: does this not work with the existing framework? @edmondo1984 - is it documented/clear/easy to add more reference tables?

also - from an API perspective - do we want users to be able to generate just a single reference table? I think so.

@edmondo1984 is there an existing API for that?

File: dat/main.py [hunk 2]
```diff
@@ -119,6 +124,7 @@ def create_external_tables(external_tables_path, output_path):
 cli.add_command(write_generated_reference_tables)
 cli.add_command(write_schemas)
 cli.add_command(create_external_tables)
+cli.add_command(write_generated_reference_table_3)
```
Comment (scottsand-db) [inline dat/main.py:127]: IMO we should rethink the "generate a single reference table" later, and just have this reference table be part of the tables generated by the current command, which generates all of them.

File: dat/writers/create_reference_tables.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
+    .config(
+        "spark.sql.catalog.spark_catalog",
+        "org.apache.spark.sql.delta.catalog.DeltaCatalog",
+    )
+)
+
+spark = configure_spark_with_delta_pip(builder).getOrCreate()
+
+def create_reference_table_3():
```
Comment (scottsand-db) [inline dat/writers/create_reference_tables.py:17]: isn't this supposed to generate the `table-metadata.json` too?

File: dat/main.py (hunk 1 above)
Comment (edmondo1984) [reply dat/main.py:72]: https://github.com/delta-incubator/dat/pull/17/files/385b3822570fcec23d7e04a577f938ed8f6ef682..83f3e7e1b4a65ad68c5ff67fb18a476c9e2dbeca#diff-43f89dd411632a233f2c80cadaf9c223cbe98f2bb6bf5cad6f74f01d811b16c3R39

if someone would specify --table-name=foo, the only table name with table foo would be specified

File: tests/pyspark_delta/test_pyspark_delta.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+builder = (
+    pyspark.sql.SparkSession.builder.appName("MyApp")
+    .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
+    .config(
+        "spark.sql.catalog.spark_catalog",
+        "org.apache.spark.sql.delta.catalog.DeltaCatalog",
+    )
+)
+
+spark = configure_spark_with_delta_pip(builder).getOrCreate()
```
Comment (edmondo1984) [inline tests/pyspark_delta/test_pyspark_delta.py:15]: Expensive, reusable pieces of infrastructure should leverage pytest.fixtures, so that the context is created once and injected in all tests that need it.  and it's already there, by the way, here

https://github.com/delta-incubator/dat/blob/master/tests/conftest.py#L7

File: tests/pyspark_delta/test_pyspark_delta.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+def test_reference_table_1_content():
```
Comment (edmondo1984) [inline tests/pyspark_delta/test_pyspark_delta.py:17]:
```suggestion
def test_reference_table_1_content(spark):
```

**Review by edmondo1984 — CHANGES_REQUESTED** [review]

File: dat/writers/create_reference_tables.py [hunk 6]
```diff
@@ -14,6 +14,54 @@
 
 spark = configure_spark_with_delta_pip(builder).getOrCreate()
 
+def create_reference_table_1():
+    columns = ["letter", "number", "a_float"]
```
Comment (edmondo1984) [inline dat/writers/create_reference_tables.py:18]: These functions violates the single responsibility principle, see https://en.wikipedia.org/wiki/Single-responsibility_principle or "Clean Code" from Uncle Bob.

These function performs: 

1. creating a first dataframe
2. creating path on the file system
3. writing the first dataframe to disk
4. writing the dataframe content as parquet
5. creating a second dataframe
6. writing the second dataframe to disk

First of all, the second activity could be performed as a first one, so the function looks like that

```python

def _ensure_path_exists(table_name):
        table_path =f 'out/tables/generated/{table_name}/expected'
        os.makedirs(table_path)

def create_reference_table_3():
       _ensure_path_exist('reference_table_3')
      df_1 = _create_df_1()
      write_df_as_delta_and_spark(df_1)
      df_2 = _create_df_2()
      write_df_as_delta_and_spark(df_2)
```

Comment (edmondo1984) [reply dat/writers/create_reference_tables.py:18]: Now if you take this approach, and you generalize it with respect to the number of transactions, you use an array

```python
def create_reference_table_3():
       _ensure_path_exist('reference_table_3')
      dfs = [ _create_df_1(), _create_df_2() ]
      for df in dfs:
             write_df_as_delta_and_spark(df)
```

If you keep iterating by applying good coding practices such as DRY, single responsibility principle, etc, you end up with something similar to the existing codebase

**Review by edmondo1984 — CHANGES_REQUESTED** [review]

File: tests/pyspark_delta/test_pyspark_delta.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    for v in versions:
+        actual_df = spark.read.format("delta").option("versionAsOf", v).load("out/tables/generated/reference_table_2/delta")
+        expected_df = spark.read.format("parquet").load(f"out/tables/generated/reference_table_2/expected/table_content_v{v}.parquet")
+        chispa.assert_df_equality(actual_df, expected_df)
+
+
+def test_reference_table_3_content():
+    versions = [0, 1, 2]
+    for v in versions:
+        actual_df = spark.read.format("delta").option("versionAsOf", v).load("out/tables/generated/reference_table_3/delta")
```
[the remark this answers is not in the document]
Comment (edmondo1984) [reply tests/pyspark_delta/test_pyspark_delta.py:36]: (I made a typo, the name of the arg should be v and not version to match the code)

File: dat/writers/create_reference_tables.py (hunk 6 above)
Comment (MrPowers) [reply dat/writers/create_reference_tables.py:18]: @edmondo1984 - yea, we'll definitely want to make abstractions and clean the code up.  I'm just not sure what the right abstractions are yet.  I wasn't sure how to fit `repartition(1)`, alter table properties, etc. into the current framework.  I am going to try and make a more diverse list of reference tables and then see if we can make abstractions that are sufficiently abstract to handle all the edge cases.

File: tests/pyspark_delta/test_pyspark_delta.py (hunk 4 above)
Comment (MrPowers) [reply tests/pyspark_delta/test_pyspark_delta.py:15]: Yea, what you have looks better, but can we take out the `spark.stop` so the tests run faster?

File: tests/pyspark_delta/test_pyspark_delta.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        chispa.assert_df_equality(actual_df, expected_df)
+
+
+def test_all_readers():
+    test_config = [
+        {"rt": 1, "versions": [0, 1]},
+        {"rt": 2, "versions": [0, 1]},
+        {"rt": 3, "versions": [0, 1, 2]},
+    ]
+    for config in test_config:
```
[the remark this answers is not in the document]
Comment (MrPowers) [reply tests/pyspark_delta/test_pyspark_delta.py:47]: Yea, loops are definitely a test antipattern, but think they might be useful here.  We're trying to make it really easy for connectors to write hundreds of tests.  We need to find the right interface to provide connectors so they can write all these tests without specifying each one manually.  All of this code will be thrown out.  It's just hopefully getting us closer to seeing what's the right abstraction.

File: dat/writers/create_reference_tables.py (hunk 6 above)
Comment (edmondo1984) [reply dat/writers/create_reference_tables.py:18]: I see, yes. I guess one has to extend the object model (there is a RowCollection), you probably need to change that into RowOperation with multiple subclasses (inserts, repertitions)

File: tests/pyspark_delta/test_pyspark_delta.py (hunk 4 above)
Comment (edmondo1984) [reply tests/pyspark_delta/test_pyspark_delta.py:15]: the finalizer is another fixture that is executed by pytest when all tests are run, see https://docs.pytest.org/en/6.2.x/reference.html#std-fixture-request

File: dat/writers/create_reference_tables.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    data = [("d", 4, 4.4), ("e", 5, 5.5)]
+    rdd = spark.sparkContext.parallelize(data)
+    df = rdd.toDF(columns)    
+
+    df.repartition(1).write.format("delta").mode("append").save(delta_table_path)
+    actual1 = spark.read.format("delta").load(delta_table_path)
+    actual1.toPandas().to_parquet(f"{expected_path}/table_content_v1.parquet")
+
+
+def create_reference_table_2():
```
Comment (wjones127) [inline dat/writers/create_reference_tables.py:41]: It would be cool to have some human-readable description for each of these, so we have a general idea of what the case is testing.

File: tests/pyspark_delta/test_pyspark_delta.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+def test_reference_table_1_content():
+    versions = [0, 1]
+    for v in versions:
+        actual_df = spark.read.format("delta").option("versionAsOf", v).load("out/tables/generated/reference_table_1/delta")
+        expected_df = spark.read.format("parquet").load(f"out/tables/generated/reference_table_1/expected/v{v}/table_content.parquet")
+        chispa.assert_df_equality(actual_df, expected_df)
+
+
+def test_reference_table_2_content():
```
Comment (wjones127) [inline tests/pyspark_delta/test_pyspark_delta.py:25]: Instead of writing a test function for each case, we should be able to write code that will collect all the test cases and them pass them into one parameterized case. See what I did in: https://github.com/delta-io/delta-rs/pull/909/files

**Review by wjones127 — APPROVED** [review]
